spatialprofilingtoolbox/ondemand/providers/pending_provider.py

```diff
@@ -0,0 +1,194 @@
+"""Abstract class for providers that need to wait for on demand calculations to complete."""
+
+from abc import ABC, abstractmethod
+from threading import Thread
+from datetime import datetime
+
+from pandas import DataFrame
+
+from spatialprofilingtoolbox.db.database_connection import DBCursor
+from spatialprofilingtoolbox.db.exchange_data_formats.metrics import PhenotypeCriteria
+from spatialprofilingtoolbox.ondemand.providers import Provider
+from spatialprofilingtoolbox.standalone_utilities.log_formats import colorized_logger
+
+logger = colorized_logger(__name__)
+
+
+class PendingProvider(Provider, ABC):
+    """Provide precalculated metrics or start, wait, receive, and then provide them."""
+
+    def get_metrics(
+        self,
+        study_name: str,
+        **kwargs: int | PhenotypeCriteria | list[PhenotypeCriteria]
+    ) -> dict[str, dict[str, float | None] | bool]:
+        """Get requested metrics or signal that it's not done calculating yet."""
+        logger.debug('Requesting computation.')
+        feature_specification = self.get_or_create_feature_specification(
+            study_name, **kwargs)
+        if self._is_already_computed(feature_specification):
+            is_pending = False
+            logger.debug('Already computed.')
+        else:
+            is_pending = self._is_already_pending(feature_specification)
+            if is_pending:
+                logger.debug('Already already pending.')
+            else:
+                logger.debug('Not already pending.')
+            if not is_pending:
+                logger.debug('Starting background task.')
+                self._fork_computation_task(feature_specification)
+                self._set_pending_computation(feature_specification)
+                logger.debug('Background task just started, is pending.')
+                is_pending = True
+        return self._query_for_computed_feature_values(
+            feature_specification, still_pending=is_pending)
+
+    @abstractmethod
+    @classmethod
+    def get_or_create_feature_specification(
+        cls,
+        study_name: str,
+        **kwargs: int | PhenotypeCriteria | list[PhenotypeCriteria]
```

Comment:
The specifiers for a feature specification can theoretically be quite arbitrary. The explicit list of types here are just parroting back the 3 cases we currently have and are generalizing: radii, a phenotype, or multiple phenotypes. But the `scstudies` schema allows that the specifiers could be arbitrary strings, whose meaning is supposed to be determinable by reading the sentential "feature derivation method" text. I'm thinking about possible specifiers like "**mean** or** median**" in case the feature derivation method involves a statistical center value. I think it is misleading to indicate that these specifiers will always be cell phenotypes or a single numerical quantity.

Reply:
So I guess I am advocating `specifiers: tuple[str]`.

Reply:
I seem to be leading us down a road toward a certain conflict, between:

1. The notion that features that are cell phenotypes should be specifiable on-demand (the so-called anonymous phenotypes, passed by value, as a list of positives and and list of negatives).
2. The notion that feature specifications (in order for the values to be stored correctly in the DB) should have a definite fixed-length list of specifier strings.

The conflict is resolveable easily if part of the on demand job is to insert new phenotypes sometimes, with internal-use-only names that are just integers or something similar.

That proposal raises some questions for me though, about possibly giving too much record-insertion power to arbitrary API users. Maybe it is not such a big deal, after all we are inserting computed values on request.

Reply:
The important things that we need to keep track of are
* what source data columns are we using (radii and phenotypes, in this case)
* how we transformed them (this is what I think the feature specification is/should be)
* what that transformed data is
for the sake of replicability. (To that end we may also want to track the random seed as well lol.)

`tuple[str]` is fine but if we're going to do that, at least for `SquidpyProvider` and `PhenotypeProvider`, we should extend the function such that `PhenotypeCriteria` can get involved in the typing and use the controlled criteria_to_string function to make it maximally clear what should be done for future Providers. The feature specification should be as formulaic as possible if we're going to let it be a bunch of strings.

Reply:
Note: I wouldn't call radii/phenotypes in this example "source data columns". They are "inputs" but more in the sense of being parameters for a function (namely the feature derivation function), not first-class source data values. (The real source data is anything from the study's original DB instance; cell locations, channel values, channel metadata for the antibodies, diagnosis, intervention, specimen types... .)

Again, I also wouldn't use the term "transformed" here because "transformed" connotes a more general concept than we want. E.g. it allows 5000-entry matrices as the transformed values. We don't want just any type of transformed data to get recorded here, only "features". The verb I've been using is "derive" as in "feature derivation method".

Yes I agree that `PhenotypeCriteria` or similar should be involved somehow in the typing in the implementation of the providers. I think I'm arguing that it should not be involved in the provider interface however, since the exact types are completely feature-type-dependent.
On the other hand the provider interface *should* indicate how to do the object-to-string, as you say.
Perhaps we need `specifiers: tuple[FeatureSpecifier]` and for `PhenotypeCriteria` to implement a `FeatureSpecifier` interface with `to_string`.